{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PyRosetta-4 2022 [Rosetta PyRosetta4.conda.linux.cxx11thread.serialization.CentOS.python39.Release 2022.04+release.0c9f888efdecfed435625915cc2c8cddbb485ee9 2022-01-26T19:42:42] retrieved from: http://www.pyrosetta.org\n",
      "(C) Copyright Rosetta Commons Member Institutions. Created in JHU by Sergey Lyskov and PyRosetta Team.\n",
      "core.init: Checking for fconfig files in pwd and ./rosetta/flags\n",
      "core.init: Rosetta version: PyRosetta4.conda.linux.cxx11thread.serialization.CentOS.python39.Release r309 2022.04+release.0c9f888 0c9f888efdecfed435625915cc2c8cddbb485ee9 http://www.pyrosetta.org 2022-01-26T19:42:42\n",
      "core.init: command: PyRosetta -ex1 -ex2aro -database /home/tsatler/anaconda3/envs/pyro/lib/python3.9/site-packages/pyrosetta/database\n",
      "basic.random.init_random_generator: 'RNG device' seed mode, using '/dev/urandom', seed=1587666856 seed_offset=0 real_seed=1587666856 thread_index=0\n",
      "basic.random.init_random_generator: RandomGenerator:init: Normal mode, seed=1587666856 RG_type=mt19937\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import glob\n",
    "import json\n",
    "import shutil\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from pyrosetta import *\n",
    "\n",
    "init()\n",
    "from pyrosetta.rosetta.protocols import *\n",
    "from pyrosetta.rosetta.protocols.rosetta_scripts import XmlObjects\n",
    "from Bio.PDB.Polypeptide import aa1, aa3\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "from Bio.PDB import PDBParser"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_scores(scores_path, binder_len=None, is_binder_second=False):\n",
    "    scores = json.loads(Path(scores_path).read_text())\n",
    "\n",
    "    plddt = np.mean(scores[\"plddt\"])\n",
    "    pae = np.array(scores[\"pae\"])\n",
    "\n",
    "    if is_binder_second:\n",
    "        pae_binder = np.mean(pae[binder_len:, binder_len:]) if binder_len else None\n",
    "        pae_target = np.mean(pae[:binder_len, :binder_len]) if binder_len else None\n",
    "        plddt_binder = np.mean(scores[\"plddt\"][binder_len:]) if binder_len else None\n",
    "        plddt_target = np.mean(scores[\"plddt\"][:binder_len]) if binder_len else None\n",
    "        pae_int1 = np.mean(pae[binder_len:, :binder_len]) if binder_len else None\n",
    "        pae_int2 = np.mean(pae[:binder_len, binder_len:]) if binder_len else None\n",
    "    else:\n",
    "        pae_binder = np.mean(pae[:binder_len, :binder_len]) if binder_len else None\n",
    "        pae_target = np.mean(pae[binder_len:, binder_len:]) if binder_len else None\n",
    "        plddt_binder = np.mean(scores[\"plddt\"][:binder_len]) if binder_len else None\n",
    "        plddt_target = np.mean(scores[\"plddt\"][binder_len:]) if binder_len else None\n",
    "        pae_int1 = np.mean(pae[:binder_len, binder_len:]) if binder_len else None\n",
    "        pae_int2 = np.mean(pae[binder_len:, :binder_len]) if binder_len else None\n",
    "\n",
    "    pae_int_tot = (pae_int1 + pae_int2) / 2 if binder_len else None\n",
    "\n",
    "    results = {\"plddt\": plddt, \"pae\": np.mean(pae)}\n",
    "    if binder_len:\n",
    "        results.update(\n",
    "            {\n",
    "                \"binder_plddt\": plddt_binder,\n",
    "                \"target_plddt\": plddt_target,\n",
    "                \"pae_binder\": pae_binder,\n",
    "                \"pae_target\": pae_target,\n",
    "                \"pae_int_tot\": pae_int_tot,\n",
    "            }\n",
    "        )\n",
    "\n",
    "    return results\n",
    "\n",
    "\n",
    "def calculate_residue_difference(seq1, seq2):\n",
    "    # Ensure sequences have the same length\n",
    "    if len(seq1) != len(seq2):\n",
    "        raise ValueError(\"Sequences must have the same length.\")\n",
    "\n",
    "    # Calculate the number of different residues\n",
    "    num_different = sum(a != b for a, b in zip(seq1, seq2))\n",
    "\n",
    "    # Calculate the percentage of different residues\n",
    "    percentage = (num_different / len(seq1)) * 100\n",
    "\n",
    "    return percentage\n",
    "\n",
    "\n",
    "def align_structures(pdb1, pdb2):\n",
    "    \"\"\"Take two structure and superimpose pdb1 on pdb2\"\"\"\n",
    "    import Bio.PDB\n",
    "\n",
    "    pdb_parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    # Get the structures\n",
    "    ref_structure = pdb_parser.get_structure(\"ref\", pdb1)\n",
    "    sample_structure = pdb_parser.get_structure(\"sample\", pdb2)\n",
    "\n",
    "    aligner = Bio.PDB.cealign.CEAligner()\n",
    "    aligner.set_reference(ref_structure)\n",
    "    aligner.align(sample_structure)\n",
    "\n",
    "    return aligner.rms\n",
    "\n",
    "\n",
    "def get_sasa(pose):\n",
    "    \"\"\"Calculate the total and hydrophobic sasa\"\"\"\n",
    "    rsd_sasa = pyrosetta.rosetta.utility.vector1_double()\n",
    "    rsd_hydrophobic_sasa = pyrosetta.rosetta.utility.vector1_double()\n",
    "    rosetta.core.scoring.calc_per_res_hydrophobic_sasa(\n",
    "        pose, rsd_sasa, rsd_hydrophobic_sasa, 1.4\n",
    "    )  # The last arguement is the probe radius\n",
    "\n",
    "    return sum(rsd_sasa), sum(rsd_hydrophobic_sasa)\n",
    "\n",
    "\n",
    "def calculate_charge(chain, ph=7.4):\n",
    "    \"\"\"\n",
    "    Calculates the charge of the protein chain at a specific pH.\n",
    "\n",
    "    Parameters:\n",
    "        chain (Bio.PDB.Chain.Chain): The protein chain to calculate the charge of.\n",
    "        ph (float): The pH value to calculate the charge at. Default is 7.4.\n",
    "\n",
    "    Returns:\n",
    "        The charge of the protein at the specified pH.\n",
    "    \"\"\"\n",
    "    # Extract the sequence of amino acid residues in the chain\n",
    "    sequence = \"\"\n",
    "    for residue in chain.get_residues():\n",
    "        resname = residue.get_resname()\n",
    "        if resname in aa3:\n",
    "            sequence += aa1[aa3.index(resname)]\n",
    "        else:\n",
    "            print(\n",
    "                f\"Skipping residue {resname} because it is not a standard amino acid.\"\n",
    "            )\n",
    "            continue\n",
    "\n",
    "    # Create a ProteinAnalysis object from the sequence\n",
    "    protein_analysis = ProteinAnalysis(sequence)\n",
    "\n",
    "    # Calculate the charge of the protein at a specific pH\n",
    "    charge = protein_analysis.charge_at_pH(ph)\n",
    "\n",
    "    return charge\n",
    "\n",
    "\n",
    "def calculate_sap_score(pose, chain=\"B\"):\n",
    "\n",
    "    # Select only chain B using a SwitchChainOrder mover\n",
    "    select_chain = XmlObjects.static_get_mover(\n",
    "        f'<SwitchChainOrder name=\"so\" chain_order=\"{chain}\"/>'\n",
    "    )\n",
    "    chain = pose.clone()\n",
    "    select_chain.apply(chain)\n",
    "\n",
    "    # Calculate the SAP score for chain B\n",
    "    sap_score_metric = XmlObjects.static_get_simple_metric(\n",
    "        '<SapScoreMetric name=\"sap_metric\"/>'\n",
    "    )\n",
    "    sap_score_value = sap_score_metric.calculate(chain)\n",
    "\n",
    "    # Return the SAP score value\n",
    "    # sap_score_value = sap_score_metric.get(1)\n",
    "    return sap_score_value"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0_1_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_2_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_3_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_4_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_5_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_6_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_7_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_8_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_9_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_10_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_11_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_12_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_13_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_14_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_15_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_16_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_17_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_18_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_19_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_20_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_21_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_22_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_23_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_24_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_25_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_26_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_27_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_28_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_29_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "0_30_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_31_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_32_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_33_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_34_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_35_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_36_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_37_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_38_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_39_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_40_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_41_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_42_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_43_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_44_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_45_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_46_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_47_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "1_48_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "core.chemical.GlobalResidueTypeSet: Finished initializing fa_standard residue type set.  Created 983 residue types\n",
      "core.chemical.GlobalResidueTypeSet: Total time to initialize 1.17435 seconds.\n",
      "core.import_pose.import_pose: File 'output/7urv_FMC63_complex/tmp/af2_in/AF2out_21/1_49_7urv_FMC63_complex_unrelaxed_rank_001_alphafold2_multimer_v3_model_4_seed_000.pdb' automatically determined to be of type PDB\n",
      "core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue SER:CtermProteinFull 242\n",
      "core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue ARG:CtermProteinFull 460\n",
      "core.conformation.Conformation: Found disulfide between residues 23 88\n",
      "core.conformation.Conformation: current variant for 23 CYS\n",
      "core.conformation.Conformation: current variant for 88 CYS\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 258 444\n",
      "core.conformation.Conformation: current variant for 258 CYS\n",
      "core.conformation.Conformation: current variant for 444 CYS\n",
      "core.conformation.Conformation: current variant for 258 CYD\n",
      "core.conformation.Conformation: current variant for 444 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 310 383\n",
      "core.conformation.Conformation: current variant for 310 CYS\n",
      "core.conformation.Conformation: current variant for 383 CYS\n",
      "core.conformation.Conformation: current variant for 310 CYD\n",
      "core.conformation.Conformation: current variant for 383 CYD\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Validating input script...\n",
      "protocols.rosetta_scripts.RosettaScriptsSchemaValidator: Generating XML Schema for rosetta_scripts...\n",
      "protocols.rosetta_scripts.RosettaScriptsSchemaValidator: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsSchemaValidator: Initializing schema validator...\n",
      "protocols.rosetta_scripts.RosettaScriptsSchemaValidator: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Parsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<MOVERS>\n",
      "\t\t<SwitchChainOrder chain_order=\"A\" name=\"so\"/>\n",
      "\t</MOVERS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "core.scoring.etable: Starting energy table calculation\n",
      "core.scoring.etable: smooth_etable: changing atr/rep split to bottom of energy well\n",
      "core.scoring.etable: smooth_etable: spline smoothing lj etables (maxdis = 6)\n",
      "core.scoring.etable: smooth_etable: spline smoothing solvation etables (max_dis = 6)\n",
      "core.scoring.etable: Finished calculating energy tables.\n",
      "basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/HBPoly1D.csv\n",
      "basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/HBFadeIntervals.csv\n",
      "basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/HBEval.csv\n",
      "basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/DonStrength.csv\n",
      "basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/AccStrength.csv\n",
      "basic.io.database: Database file opened: scoring/score_functions/rama/fd/all.ramaProb\n",
      "basic.io.database: Database file opened: scoring/score_functions/rama/fd/prepro.ramaProb\n",
      "basic.io.database: Database file opened: scoring/score_functions/omega/omega_ppdep.all.txt\n",
      "basic.io.database: Database file opened: scoring/score_functions/omega/omega_ppdep.gly.txt\n",
      "basic.io.database: Database file opened: scoring/score_functions/omega/omega_ppdep.pro.txt\n",
      "basic.io.database: Database file opened: scoring/score_functions/omega/omega_ppdep.valile.txt\n",
      "basic.io.database: Database file opened: scoring/score_functions/P_AA_pp/P_AA\n",
      "basic.io.database: Database file opened: scoring/score_functions/P_AA_pp/P_AA_n\n",
      "core.scoring.P_AA: shapovalov_lib::shap_p_aa_pp_smooth_level of 1( aka low_smooth ) got activated.\n",
      "basic.io.database: Database file opened: scoring/score_functions/P_AA_pp/shapovalov/10deg/kappa131/a20.prop\n",
      "core.scoring.etable: Starting energy table calculation\n",
      "core.scoring.etable: smooth_etable: changing atr/rep split to bottom of energy well\n",
      "core.scoring.etable: smooth_etable: spline smoothing lj etables (maxdis = 6)\n",
      "core.scoring.etable: smooth_etable: spline smoothing solvation etables (max_dis = 6)\n",
      "core.scoring.etable: Finished calculating energy tables.\n",
      "basic.io.database: Database file opened: scoring/score_functions/PairEPotential/pdb_pair_stats_fine\n",
      "basic.io.database: Database file opened: scoring/score_functions/InterchainPotential/interchain_env_log.txt\n",
      "basic.io.database: Database file opened: scoring/score_functions/InterchainPotential/interchain_pair_log.txt\n",
      "basic.io.database: Database file opened: scoring/score_functions/EnvPairPotential/env_log.txt\n",
      "basic.io.database: Database file opened: scoring/score_functions/EnvPairPotential/cbeta_den.txt\n",
      "basic.io.database: Database file opened: scoring/score_functions/EnvPairPotential/pair_log.txt\n",
      "basic.io.database: Database file opened: scoring/score_functions/EnvPairPotential/cenpack_log.txt\n",
      "core.scoring.ramachandran: shapovalov_lib::shap_rama_smooth_level of 4( aka highest_smooth ) got activated.\n",
      "basic.io.database: Database file opened: scoring/score_functions/rama/shapovalov/kappa25/all.ramaProb\n",
      "basic.io.database: Database file opened: scoring/score_functions/rama/flat/avg_L_rama.dat\n",
      "core.scoring.ramachandran: Reading custom Ramachandran table from scoring/score_functions/rama/flat/avg_L_rama.dat.\n",
      "basic.io.database: Database file opened: scoring/score_functions/rama/flat/sym_all_rama.dat\n",
      "core.scoring.ramachandran: Reading custom Ramachandran table from scoring/score_functions/rama/flat/sym_all_rama.dat.\n",
      "basic.io.database: Database file opened: scoring/score_functions/rama/flat/sym_G_rama.dat\n",
      "core.scoring.ramachandran: Reading custom Ramachandran table from scoring/score_functions/rama/flat/sym_G_rama.dat.\n",
      "basic.io.database: Database file opened: scoring/score_functions/rama/flat/sym_P_rama.dat\n",
      "core.scoring.ramachandran: Reading custom Ramachandran table from scoring/score_functions/rama/flat/sym_P_rama.dat.\n",
      "basic.io.database: Database file opened: scoring/score_functions/rama/flat/avg_L_rama_str.dat\n",
      "core.scoring.ramachandran: Reading custom Ramachandran table from scoring/score_functions/rama/flat/avg_L_rama_str.dat.\n",
      "basic.io.database: Database file opened: scoring/score_functions/rama/flat/sym_all_rama_str.dat\n",
      "core.scoring.ramachandran: Reading custom Ramachandran table from scoring/score_functions/rama/flat/sym_all_rama_str.dat.\n",
      "basic.io.database: Database file opened: scoring/score_functions/rama/flat/sym_G_rama_str.dat\n",
      "core.scoring.ramachandran: Reading custom Ramachandran table from scoring/score_functions/rama/flat/sym_G_rama_str.dat.\n",
      "basic.io.database: Database file opened: scoring/score_functions/rama/flat/sym_P_rama_str.dat\n",
      "core.scoring.ramachandran: Reading custom Ramachandran table from scoring/score_functions/rama/flat/sym_P_rama_str.dat.\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Defined mover named \"so\" of type SwitchChainOrder\n",
      "protocols.rosetta_scripts.ParsedProtocol: ParsedProtocol mover with the following settings\n",
      "protocols.simple_moves.SwitchChainOrderMover: Number of chains in pose: 2\n",
      "protocols.simple_moves.SwitchChainOrderMover: Now at chain: 1\n",
      "core.conformation.Conformation: Found disulfide between residues 23 88\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "basic.io.database: Database file opened: scoring/score_functions/elec_cp_reps.dat\n",
      "core.scoring.elec.util: Read 40 countpair representative atoms\n",
      "core.pack.dunbrack.RotamerLibrary: shapovalov_lib_fixes_enable option is true.\n",
      "core.pack.dunbrack.RotamerLibrary: shapovalov_lib::shap_dun10_smooth_level of 1( aka lowest_smooth ) got activated.\n",
      "core.pack.dunbrack.RotamerLibrary: Binary rotamer library selected: /home/tsatler/anaconda3/envs/pyro/lib/python3.9/site-packages/pyrosetta/database/rotamer/shapovalov/StpDwn_0-0-0/Dunbrack10.lib.bin\n",
      "core.pack.dunbrack.RotamerLibrary: Using Dunbrack library binary file '/home/tsatler/anaconda3/envs/pyro/lib/python3.9/site-packages/pyrosetta/database/rotamer/shapovalov/StpDwn_0-0-0/Dunbrack10.lib.bin'.\n",
      "core.pack.dunbrack.RotamerLibrary: Dunbrack 2010 library took 0.248752 seconds to load from binary\n",
      "protocols.simple_moves.SwitchChainOrderMover: New pose's foldtree FOLD_TREE  EDGE 1 242 -1\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Validating input script...\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Parsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<SIMPLE_METRICS>\n",
      "\t\t<SapScoreMetric name=\"sap_metric\"/>\n",
      "\t</SIMPLE_METRICS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "protocols.rosetta_scripts.ParsedProtocol: ParsedProtocol mover with the following settings\n",
      "basic.io.database: Database file opened: scoring/score_functions/sap_sasa_calib.dat\n",
      "core.import_pose.import_pose: File 'output/7urv_FMC63_complex/tmp/af2_in/AF2out_21/1_50_7urv_FMC63_complex_unrelaxed_rank_001_alphafold2_multimer_v3_model_2_seed_000.pdb' automatically determined to be of type PDB\n",
      "core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue SER:CtermProteinFull 242\n",
      "core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue ARG:CtermProteinFull 460\n",
      "core.conformation.Conformation: Found disulfide between residues 23 88\n",
      "core.conformation.Conformation: current variant for 23 CYS\n",
      "core.conformation.Conformation: current variant for 88 CYS\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 258 444\n",
      "core.conformation.Conformation: current variant for 258 CYS\n",
      "core.conformation.Conformation: current variant for 444 CYS\n",
      "core.conformation.Conformation: current variant for 258 CYD\n",
      "core.conformation.Conformation: current variant for 444 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 310 383\n",
      "core.conformation.Conformation: current variant for 310 CYS\n",
      "core.conformation.Conformation: current variant for 383 CYS\n",
      "core.conformation.Conformation: current variant for 310 CYD\n",
      "core.conformation.Conformation: current variant for 383 CYD\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Validating input script...\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Parsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<MOVERS>\n",
      "\t\t<SwitchChainOrder chain_order=\"A\" name=\"so\"/>\n",
      "\t</MOVERS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Defined mover named \"so\" of type SwitchChainOrder\n",
      "protocols.rosetta_scripts.ParsedProtocol: ParsedProtocol mover with the following settings\n",
      "protocols.simple_moves.SwitchChainOrderMover: Number of chains in pose: 2\n",
      "protocols.simple_moves.SwitchChainOrderMover: Now at chain: 1\n",
      "core.conformation.Conformation: Found disulfide between residues 23 88\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "protocols.simple_moves.SwitchChainOrderMover: New pose's foldtree FOLD_TREE  EDGE 1 242 -1\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Validating input script...\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Parsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<SIMPLE_METRICS>\n",
      "\t\t<SapScoreMetric name=\"sap_metric\"/>\n",
      "\t</SIMPLE_METRICS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "protocols.rosetta_scripts.ParsedProtocol: ParsedProtocol mover with the following settings\n",
      "core.import_pose.import_pose: File 'output/7urv_FMC63_complex/tmp/af2_in/AF2out_21/1_51_7urv_FMC63_complex_unrelaxed_rank_001_alphafold2_multimer_v3_model_2_seed_000.pdb' automatically determined to be of type PDB\n",
      "core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue LEU:CtermProteinFull 242\n",
      "core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue ARG:CtermProteinFull 460\n",
      "core.conformation.Conformation: Found disulfide between residues 23 88\n",
      "core.conformation.Conformation: current variant for 23 CYS\n",
      "core.conformation.Conformation: current variant for 88 CYS\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 258 444\n",
      "core.conformation.Conformation: current variant for 258 CYS\n",
      "core.conformation.Conformation: current variant for 444 CYS\n",
      "core.conformation.Conformation: current variant for 258 CYD\n",
      "core.conformation.Conformation: current variant for 444 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 310 383\n",
      "core.conformation.Conformation: current variant for 310 CYS\n",
      "core.conformation.Conformation: current variant for 383 CYS\n",
      "core.conformation.Conformation: current variant for 310 CYD\n",
      "core.conformation.Conformation: current variant for 383 CYD\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Validating input script...\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Parsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<MOVERS>\n",
      "\t\t<SwitchChainOrder chain_order=\"A\" name=\"so\"/>\n",
      "\t</MOVERS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Defined mover named \"so\" of type SwitchChainOrder\n",
      "protocols.rosetta_scripts.ParsedProtocol: ParsedProtocol mover with the following settings\n",
      "protocols.simple_moves.SwitchChainOrderMover: Number of chains in pose: 2\n",
      "protocols.simple_moves.SwitchChainOrderMover: Now at chain: 1\n",
      "core.conformation.Conformation: Found disulfide between residues 23 88\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "protocols.simple_moves.SwitchChainOrderMover: New pose's foldtree FOLD_TREE  EDGE 1 242 -1\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Validating input script...\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Parsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<SIMPLE_METRICS>\n",
      "\t\t<SapScoreMetric name=\"sap_metric\"/>\n",
      "\t</SIMPLE_METRICS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "protocols.rosetta_scripts.ParsedProtocol: ParsedProtocol mover with the following settings\n",
      "core.import_pose.import_pose: File 'output/7urv_FMC63_complex/tmp/af2_in/AF2out_21/1_52_7urv_FMC63_complex_unrelaxed_rank_001_alphafold2_multimer_v3_model_1_seed_000.pdb' automatically determined to be of type PDB\n",
      "core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue SER:CtermProteinFull 242\n",
      "core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue ARG:CtermProteinFull 460\n",
      "core.conformation.Conformation: Found disulfide between residues 23 88\n",
      "core.conformation.Conformation: current variant for 23 CYS\n",
      "core.conformation.Conformation: current variant for 88 CYS\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 144 217\n",
      "core.conformation.Conformation: current variant for 144 CYS\n",
      "core.conformation.Conformation: current variant for 217 CYS\n",
      "core.conformation.Conformation: current variant for 144 CYD\n",
      "core.conformation.Conformation: current variant for 217 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 258 444\n",
      "core.conformation.Conformation: current variant for 258 CYS\n",
      "core.conformation.Conformation: current variant for 444 CYS\n",
      "core.conformation.Conformation: current variant for 258 CYD\n",
      "core.conformation.Conformation: current variant for 444 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 310 383\n",
      "core.conformation.Conformation: current variant for 310 CYS\n",
      "core.conformation.Conformation: current variant for 383 CYS\n",
      "core.conformation.Conformation: current variant for 310 CYD\n",
      "core.conformation.Conformation: current variant for 383 CYD\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Validating input script...\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Parsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<MOVERS>\n",
      "\t\t<SwitchChainOrder chain_order=\"A\" name=\"so\"/>\n",
      "\t</MOVERS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Defined mover named \"so\" of type SwitchChainOrder\n",
      "protocols.rosetta_scripts.ParsedProtocol: ParsedProtocol mover with the following settings\n",
      "protocols.simple_moves.SwitchChainOrderMover: Number of chains in pose: 2\n",
      "protocols.simple_moves.SwitchChainOrderMover: Now at chain: 1\n",
      "core.conformation.Conformation: Found disulfide between residues 23 88\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 144 217\n",
      "core.conformation.Conformation: current variant for 144 CYD\n",
      "core.conformation.Conformation: current variant for 217 CYD\n",
      "core.conformation.Conformation: current variant for 144 CYD\n",
      "core.conformation.Conformation: current variant for 217 CYD\n",
      "protocols.simple_moves.SwitchChainOrderMover: New pose's foldtree FOLD_TREE  EDGE 1 242 -1\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Validating input script...\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Parsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<SIMPLE_METRICS>\n",
      "\t\t<SapScoreMetric name=\"sap_metric\"/>\n",
      "\t</SIMPLE_METRICS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "protocols.rosetta_scripts.ParsedProtocol: ParsedProtocol mover with the following settings\n",
      "core.import_pose.import_pose: File 'output/7urv_FMC63_complex/tmp/af2_in/AF2out_21/1_53_7urv_FMC63_complex_unrelaxed_rank_001_alphafold2_multimer_v3_model_3_seed_000.pdb' automatically determined to be of type PDB\n",
      "core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue ALA:CtermProteinFull 242\n",
      "core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue ARG:CtermProteinFull 460\n",
      "core.conformation.Conformation: Found disulfide between residues 23 88\n",
      "core.conformation.Conformation: current variant for 23 CYS\n",
      "core.conformation.Conformation: current variant for 88 CYS\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 144 217\n",
      "core.conformation.Conformation: current variant for 144 CYS\n",
      "core.conformation.Conformation: current variant for 217 CYS\n",
      "core.conformation.Conformation: current variant for 144 CYD\n",
      "core.conformation.Conformation: current variant for 217 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 258 444\n",
      "core.conformation.Conformation: current variant for 258 CYS\n",
      "core.conformation.Conformation: current variant for 444 CYS\n",
      "core.conformation.Conformation: current variant for 258 CYD\n",
      "core.conformation.Conformation: current variant for 444 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 310 383\n",
      "core.conformation.Conformation: current variant for 310 CYS\n",
      "core.conformation.Conformation: current variant for 383 CYS\n",
      "core.conformation.Conformation: current variant for 310 CYD\n",
      "core.conformation.Conformation: current variant for 383 CYD\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Validating input script...\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Parsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<MOVERS>\n",
      "\t\t<SwitchChainOrder chain_order=\"A\" name=\"so\"/>\n",
      "\t</MOVERS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Defined mover named \"so\" of type SwitchChainOrder\n",
      "protocols.rosetta_scripts.ParsedProtocol: ParsedProtocol mover with the following settings\n",
      "protocols.simple_moves.SwitchChainOrderMover: Number of chains in pose: 2\n",
      "protocols.simple_moves.SwitchChainOrderMover: Now at chain: 1\n",
      "core.conformation.Conformation: Found disulfide between residues 23 88\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 144 217\n",
      "core.conformation.Conformation: current variant for 144 CYD\n",
      "core.conformation.Conformation: current variant for 217 CYD\n",
      "core.conformation.Conformation: current variant for 144 CYD\n",
      "core.conformation.Conformation: current variant for 217 CYD\n",
      "protocols.simple_moves.SwitchChainOrderMover: New pose's foldtree FOLD_TREE  EDGE 1 242 -1\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Validating input script...\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Parsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<SIMPLE_METRICS>\n",
      "\t\t<SapScoreMetric name=\"sap_metric\"/>\n",
      "\t</SIMPLE_METRICS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "protocols.rosetta_scripts.ParsedProtocol: ParsedProtocol mover with the following settings\n",
      "core.import_pose.import_pose: File 'output/7urv_FMC63_complex/tmp/af2_in/AF2out_21/1_54_7urv_FMC63_complex_unrelaxed_rank_001_alphafold2_multimer_v3_model_3_seed_000.pdb' automatically determined to be of type PDB\n",
      "core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue SER:CtermProteinFull 242\n",
      "core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue ARG:CtermProteinFull 460\n",
      "core.conformation.Conformation: Found disulfide between residues 23 88\n",
      "core.conformation.Conformation: current variant for 23 CYS\n",
      "core.conformation.Conformation: current variant for 88 CYS\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 258 444\n",
      "core.conformation.Conformation: current variant for 258 CYS\n",
      "core.conformation.Conformation: current variant for 444 CYS\n",
      "core.conformation.Conformation: current variant for 258 CYD\n",
      "core.conformation.Conformation: current variant for 444 CYD\n",
      "core.conformation.Conformation: Found disulfide between residues 310 383\n",
      "core.conformation.Conformation: current variant for 310 CYS\n",
      "core.conformation.Conformation: current variant for 383 CYS\n",
      "core.conformation.Conformation: current variant for 310 CYD\n",
      "core.conformation.Conformation: current variant for 383 CYD\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Validating input script...\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Parsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<MOVERS>\n",
      "\t\t<SwitchChainOrder chain_order=\"A\" name=\"so\"/>\n",
      "\t</MOVERS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Defined mover named \"so\" of type SwitchChainOrder\n",
      "protocols.rosetta_scripts.ParsedProtocol: ParsedProtocol mover with the following settings\n",
      "protocols.simple_moves.SwitchChainOrderMover: Number of chains in pose: 2\n",
      "protocols.simple_moves.SwitchChainOrderMover: Now at chain: 1\n",
      "core.conformation.Conformation: Found disulfide between residues 23 88\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "core.conformation.Conformation: current variant for 23 CYD\n",
      "core.conformation.Conformation: current variant for 88 CYD\n",
      "protocols.simple_moves.SwitchChainOrderMover: New pose's foldtree FOLD_TREE  EDGE 1 242 -1\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Validating input script...\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: ...done\n",
      "protocols.rosetta_scripts.RosettaScriptsParser: Parsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<SIMPLE_METRICS>\n",
      "\t\t<SapScoreMetric name=\"sap_metric\"/>\n",
      "\t</SIMPLE_METRICS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "protocols.rosetta_scripts.ParsedProtocol: ParsedProtocol mover with the following settings\n",
      "Files not found for protein 1_55_7urv_FMC63_complex. Moving to next protein...\n",
      "Files not found for protein 1_56_7urv_FMC63_complex. Moving to next protein...\n",
      "Files not found for protein 1_57_7urv_FMC63_complex. Moving to next protein...\n",
      "Files not found for protein 1_58_7urv_FMC63_complex. Moving to next protein...\n",
      "Files not found for protein 1_59_7urv_FMC63_complex. Moving to next protein...\n",
      "Files not found for protein 1_60_7urv_FMC63_complex. Moving to next protein...\n",
      "2_61_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_62_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_63_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_64_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_65_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_66_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_67_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_68_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_69_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_70_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_71_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_72_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_73_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_74_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "2_75_7urv_FMC63_complex already present in the final CSV, skipping...\n",
      "core.import_pose.import_pose: File 'output/7urv_FMC63_complex/tmp/af2_in/AF2out_27/2_76_7urv_FMC63_complex_unrelaxed_rank_001_alphafold2_multimer_v3_model_2_seed_000.pdb' automatically determined to be of type PDB\n",
      "core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue SER:CtermProteinFull 242\n"
     ]
    }
   ],
   "source": [
    "##### Set intputs #####\n",
    "basename = \"7urv_FMC63_complex\"  # Basename in output folder\n",
    "chain_id = \"A\"  # Set binder chain or chain of interest for calculation of charge/sap...\n",
    "ref_seq = \"DIQMTQTTSSLSASLGDRVTISCRASQDISKYLNWYQQKPDGTVKLLIYHTSRLHSGVPSRFSGSGSGTDYSLTISNLEQEDIATYFCQQGNTLPYTFGGGTKLEITGGGGSGGGGSGGGGSEVKLQESGPGLVAPSQSLSVTCTVSGVSLPDYGVSWIRQPPRKGLEWLGVIWGSETTYYNSALKSRLTIIKDNSKSQVFLKMNSLQTDDTAIYYCAKHYYYGGSYAMDYWGQGTSVTVSS\"\n",
    "ref_pdb = \"examples/7urv_FMC63_complex.pdb\"  # Set model for RMSD calculation\n",
    "target_seq = True  # Set to True if target and binder were both predicted\n",
    "\n",
    "########################\n",
    "\n",
    "af2_input_folder = (\n",
    "    f\"output/{basename}/tmp/af2_in\"  # Folder with subfolders with AF2 predictions\n",
    ")\n",
    "af2_output_folder = f\"output/{basename}/af2\"  # Folder from 02 analysis\n",
    "af_ids = pd.read_csv(f\"{af2_output_folder}/af2_sequences.csv\")\n",
    "csv_file = f\"{af2_output_folder}/metrics.csv\"\n",
    "\n",
    "\n",
    "# Check if the CSV file already exists\n",
    "if os.path.exists(csv_file):\n",
    "    # Load the existing DataFrame from the CSV file\n",
    "    scores_df = pd.read_csv(csv_file)\n",
    "else:\n",
    "    # Create an empty DataFrame if the CSV file doesn't exist\n",
    "    if target_seq:\n",
    "        scores_df = pd.DataFrame(\n",
    "            columns=[\n",
    "                \"id\",\n",
    "                \"plddt\",\n",
    "                \"pae\",\n",
    "                \"binder_plddt\",\n",
    "                \"target_plddt\",\n",
    "                \"pae_binder\",\n",
    "                \"pae_target\",\n",
    "                \"pae_int_tot\",\n",
    "                \"diff%\",\n",
    "                \"sasa\",\n",
    "                \"hydro_sasa\",\n",
    "                \"charge\",\n",
    "                \"sap\",\n",
    "                \"rmsd\",\n",
    "                \"pdb\",\n",
    "                \"seq\",\n",
    "            ]\n",
    "        )\n",
    "    else:\n",
    "        scores_df = pd.DataFrame(\n",
    "            columns=[\n",
    "                \"id\",\n",
    "                \"plddt\",\n",
    "                \"pae\",\n",
    "                \"diff%\",\n",
    "                \"sasa\",\n",
    "                \"hydro_sasa\",\n",
    "                \"charge\",\n",
    "                \"sap\",\n",
    "                \"rmsd\",\n",
    "                \"pdb\",\n",
    "                \"seq\",\n",
    "            ]\n",
    "        )\n",
    "\n",
    "pdbs_path = f\"{af2_output_folder}/pdbs\"\n",
    "os.makedirs(pdbs_path, exist_ok=True)  # Create the directory if it doesn't exist\n",
    "\n",
    "failed_af2s = []\n",
    "# Iterate over dataframe of sequences\n",
    "for _, row in af_ids.iterrows():\n",
    "    id = row[\"id\"]\n",
    "    seq = row[\"seq\"]\n",
    "\n",
    "    if id in scores_df[\"id\"].values:\n",
    "        print(f\"{id} already present in the final CSV, skipping...\")\n",
    "        continue  # Skip to the next iteration if ID is already present\n",
    "\n",
    "    if target_seq:\n",
    "        binder_seq = seq.split(\":\")[0]  # binder sequence is first\n",
    "\n",
    "    # Check if value is already in dataframe\n",
    "    if id in scores_df[\"id\"].values:\n",
    "        print(f\"{id} already present, skipping...\")\n",
    "        continue  # Skip if 'id' is already present\n",
    "\n",
    "    try:\n",
    "        json_file = glob.glob(f\"{af2_input_folder}/*/{id}*rank_001*.json\")[0]\n",
    "        pdb_file = glob.glob(f\"{af2_input_folder}/*/{id}*rank_001*.pdb\")[0]\n",
    "    except IndexError:\n",
    "        failed_af2s.append(id)\n",
    "        print(f\"Files not found for protein {id}. Moving to next protein...\")\n",
    "        continue\n",
    "\n",
    "    # Copy best pdb file to output folder\n",
    "    shutil.copy(pdb_file, f\"{pdbs_path}/{id}.pdb\")\n",
    "\n",
    "    # Load PDB structure and create pose\n",
    "    pose = pose_from_file(pdb_file)\n",
    "    parser = PDBParser()\n",
    "    structure = parser.get_structure(\"protein\", pdb_file)\n",
    "    chain = structure[0][chain_id]\n",
    "\n",
    "    # Calculate metrics\n",
    "    if target_seq:\n",
    "        af2_scores = calculate_scores(json_file, binder_len=len(binder_seq))\n",
    "    else:\n",
    "        af2_scores = calculate_scores(json_file)\n",
    "\n",
    "    plddt = af2_scores[\"plddt\"]\n",
    "    pae = af2_scores[\"pae\"]\n",
    "    if target_seq:\n",
    "        binder_plddt = af2_scores[\"binder_plddt\"]\n",
    "        target_plddt = af2_scores[\"target_plddt\"]\n",
    "        pae_binder = af2_scores[\"pae_binder\"]\n",
    "        pae_target = af2_scores[\"pae_target\"]\n",
    "        pae_int_tot = af2_scores[\"pae_int_tot\"]\n",
    "        diff = calculate_residue_difference(ref_seq, binder_seq)\n",
    "    else:\n",
    "        diff = calculate_residue_difference(ref_seq, seq)\n",
    "    sasa, hyd_sasa = get_sasa(pose)\n",
    "    charge = calculate_charge(chain)\n",
    "    sap = calculate_sap_score(pose, chain=\"A\")\n",
    "    rmsd = align_structures(ref_pdb, pdb_file)\n",
    "\n",
    "    # Create a new row with the calculated metrics\n",
    "    if target_seq:\n",
    "        new_row = pd.DataFrame(\n",
    "            {\n",
    "                \"id\": [id],\n",
    "                \"plddt\": [plddt],\n",
    "                \"pae\": [pae],\n",
    "                \"binder_plddt\": [binder_plddt],\n",
    "                \"target_plddt\": [target_plddt],\n",
    "                \"pae_binder\": [pae_binder],\n",
    "                \"pae_target\": [pae_target],\n",
    "                \"pae_int_tot\": [pae_int_tot],\n",
    "                \"diff%\": [diff],\n",
    "                \"sasa\": [sasa],\n",
    "                \"hydro_sasa\": [hyd_sasa],\n",
    "                \"charge\": [charge],\n",
    "                \"sap\": [sap],\n",
    "                \"rmsd\": [rmsd],\n",
    "                \"pdb\": [f\"{pdbs_path}/{id}.pdb\"],\n",
    "                \"seq\": [seq],\n",
    "            }\n",
    "        )\n",
    "\n",
    "    else:\n",
    "        new_row = pd.DataFrame(\n",
    "            {\n",
    "                \"id\": id,\n",
    "                \"plddt\": [plddt],\n",
    "                \"pae\": [pae],\n",
    "                \"diff%\": [diff],\n",
    "                \"sasa\": [sasa],\n",
    "                \"hydro_sasa\": [hyd_sasa],\n",
    "                \"charge\": [charge],\n",
    "                \"sap\": [sap],\n",
    "                \"rmsd\": [rmsd],\n",
    "                \"pdb\": f\"{pdbs_path}/{id}.pdb\",\n",
    "                \"seq\": seq,\n",
    "            }\n",
    "        )\n",
    "\n",
    "    # Concatenate the new row to the DataFrame\n",
    "    scores_df = pd.concat([scores_df, new_row], ignore_index=True)\n",
    "\n",
    "    # Save the DataFrame to a file\n",
    "    scores_df.to_csv(f\"{af2_output_folder}/metrics.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\n",
    "    f\"Could not find {len(failed_af2s)} predictions! Try re-running the prediction for: {failed_af2s}\"\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyro",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
